[FEAT] marker PDF layout extraction job

priyankeshh opened the pull request:
## Summary

I have successfully implemented a comprehensive **Marker layout extraction system** for the extralit-server. This implementation provides optimized PDF layout detection capabilities that can identify and extract tables, figures, and text blocks without requiring full OCR processing.

## ✅ **What Was Implemented**

### **Core Features:**
1. **Async/Sync Job Support** - Full RQ job queue integration with both async and sync wrappers
2. **Multi-Element Extraction** - Robust extraction of tables, figures, and text blocks with bounding box coordinates

### **Files Created/Modified:**

#### **1. Main Job Implementation** (ocr_jobs.py)
- `async_marker_layout_job()` - Main async job function with full RQ integration
- `marker_layout_job()` - Synchronous wrapper for RQ compatibility  
- `_call_marker_layout_detection()` - Core Marker API integration using ConfigParser and PdfConverter
- Comprehensive error handling and job metadata tracking

#### **2. Enhanced Utility Functions**
- **tables.py** - Robust table extraction with multiple format support
- **figures.py** - Comprehensive figure extraction with enhanced validation
- **text.py** - Advanced text block extraction with subtype detection

### **Key Technical Features:**

#### **Marker Integration:**
- Uses `PdfConverter` with `ConfigParser` for optimal configuration
- Supports `create_model_dict()` for proper model initialization
- Handles page-range processing for large documents
- Markdown output format for reliable block detection

#### **Robust Data Processing:**
- Multiple naming convention support (bbox, coordinates, bounding_box, etc.)
- Various data format handling (list, dict, nested structures)
- Block type detection with extensive keyword matching
- Confidence score extraction with fallbacks


## 🚀 **Usage Example**

```python
# Extract layout from entire document
result = marker_layout_job("document.pdf", extract_text=True)

# Extract layout from specific pages
result = marker_layout_job("document.pdf", pages=[0, 1, 2])

# Result structure
{
    "tables": [{"page": 0, "bbox": [x1, y1, x2, y2], "score": 0.95, ...}],
    "figures": [{"page": 1, "bbox": [x1, y1, x2, y2], "caption": "...", ...}],
    "text_blocks": [{"page": 0, "bbox": [x1, y1, x2, y2], "content": "...", ...}],
    "metadata": {"source": "marker", "total_elements": 15, ...}
}
```

## 🔄 **Future Enhancement Opportunities**

- API endpoint exposure for HTTP access
- Batch processing support for multiple documents
- Performance metrics and monitoring integration
- Advanced configuration options for model tuning

priyankeshh committed:
feat: Replace mock Marker implementation with real PdfConverter API

- Completely replaced _call_marker_layout_detection with real marker-pdf integration
- Removed _get_mock_marker_output function in favor of actual API calls
- Updated utility functions to handle real Marker block types:
  - Tables: now handles 'table' and 'tablegroup' types
  - Figures: expanded to include 'picturegroup', 'figuregroup' types
  - Text: added support for 'sectionheader', 'textinlinemath', 'listitem', etc.
- Enhanced metadata extraction with polygon support
- Added marker-pdf as optional dependency in pyproject.toml
- Implements proper JSON output format with hierarchical block structure
- Maintains backward compatibility with existing job system and RQ decorators

priyankeshh committed:
cleaned up some code

priyankeshh committed:
minor issues

JonnyTran commented on extralit-server/src/extralit_server/jobs/ocr_jobs.py:
```diff
@@ -12,3 +12,230 @@
+    pages: Optional[list[int]] = None,
+    extract_text: bool = False,
+    document_id: Optional[UUID] = None,
+) -> dict[str, Any]:
```
We dont need to implement two functions since RQ jobs can use async functions.

JonnyTran replied:
Just deleted the function

JonnyTran committed:
Merge branch 'develop' into feat/marker-layout-extraction

priyankeshh commented:
I am getting this in `pdm run worker` terminal when i try to queue a rq worker with marker layout:
```
"D:\Extralit-gsoc\extralit\extralit               
-server\src\extralit_server\jobs\oc               
r_jobs.py", line 87, in                           
async_marker_layout_job                           
    layout_result =                               
_call_marker_layout_detection(str(p               
df_path), pages)                                  
                    ^^^^^^^^^^^^^^^               
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^               
^                                                 
  File                                            
"D:\Extralit-gsoc\extralit\extralit               
-server\src\extralit_server\jobs\oc               
r_jobs.py", line 213, in                          
_call_marker_layout_detection                     
    raise ImportError("Marker not                 
installed. Install with: pip                      
install marker-pdf") from e                       
ImportError: Marker not installed.                
Install with: pip install                         
marker-pdf                                        
```

JonnyTran committed:
Update dependencies in pdm.lock and pyproject.toml

- Updated existing dependencies: pillow (to >=10.1.0), huggingface-hub (to >=0.30.0), ocrmypdf (to >=16.11.0), marker-pdf (to >=1.9.2), torch (to >=2.5.0), torchvision (to >=0.20.0), and transformers (to >=4.51.0).

JonnyTran committed:
Merge branch 'develop' into feat/marker-layout-extraction

JonnyTran committed:
fixes

JonnyTran committed:
Update dependencies

- Updated `huggingface-hub` to version 0.34.0 in `pyproject.toml`.
- Downgraded `opencv-python-headless` to version 4.11.0.86 in `pyproject.toml`.
- Updated `marker-pdf` dependency to version 1.9.3 in `pyproject.toml`.

JonnyTran committed:
Refactor OCR job parameters and improve error logging

- Changed the type of `pages` parameter in `async_marker_layout_job` and `_call_marker_layout_detection` from `Optional[list[int]]` to `Optional[str]` to accept comma-separated page numbers.
- Updated the documentation to reflect the new format for the `pages` parameter.
- Enhanced error logging in the `async_marker_layout_job` to include exception traceback.
- Removed commented-out code related to job metadata updates for clarity.

JonnyTran committed:
Update OCR job output format to JSON

- Changed the output format in `_call_marker_layout_detection` from "markdown" to "json".
- Updated type hint for the result variable to include `JSONOutput`.
- Imported `JSONOutput` from the marker renderers for compatibility.

JonnyTran committed:
fix tests

JonnyTran commented on extralit-server/src/extralit_server/jobs/ocr_jobs.py:
```diff
@@ -12,3 +12,243 @@
+        raise ValueError(f"File is not a PDF: {pdf_path}")
+
+    try:
+        # Create optimized configuration for layout detection
+        config_dict = {
+            "output_format": "json",
+            "parallel_factor": 1,
+        }
+
+        if pages is not None:
+            config_dict["page_range"] = pages
+
+        config = ConfigParser(config_dict)
+        model_dict = create_model_dict()
+
+        converter = PdfConverter(
+            config=config.generate_config_dict(),
+            artifact_dict=model_dict,
+        )
+
+        # Convert PDF - this will return a Document object with detected layout
+        result: "MarkdownOutput | JSONOutput" = converter(pdf_path)  # noqa: UP037
+        print(type(result))
+        pprint(result.model_dump())
+
```
@priyankeshh In order to make `marker` not use the OCR models, you will have to edit the model_dict and config_dict here

JonnyTran commented on extralit-server/src/extralit_server/jobs/ocr_jobs.py:
```diff
@@ -12,3 +12,243 @@
+
+        # Extract layout information from the result
+        # The result should have metadata and blocks that we can process
+        layout_data = {"pages": []}
+
+        if hasattr(result, "pages") and result.pages:
+            for page_idx, page in enumerate(result.pages):
+                page_data = {"page": page_idx, "blocks": []}
+
+                # Extract blocks from the page
+                if hasattr(page, "blocks") and page.blocks:
+                    for block in page.blocks:
+                        if hasattr(block, "block_type") and hasattr(block, "bbox"):
+                            block_data = {
+                                "type": str(block.block_type).lower(),
+                                "bbox": list(block.bbox) if block.bbox else [],
+                                "id": str(getattr(block, "id", "")),
+                                "score": getattr(block, "confidence", 1.0),
+                            }
+
+                            # Add content based on block type
+                            if hasattr(block, "content"):
+                                block_data["content"] = str(block.content)
+                            elif hasattr(block, "text"):
+                                block_data["content"] = str(block.text)
+
+                            page_data["blocks"].append(block_data)
+
+                layout_data["pages"].append(page_data)
+
+        return layout_data
+
```
Parsing the layout json from marker output is currently broken, since `results` is being returned as a `MarkdownOutput` type, not `JSONOutput`. You may have to dig a bit to see how to do this, since `marker` don't really have documentations for it

JonnyTran commented on extralit-server/src/extralit_server/jobs/ocr_jobs.py:
```diff
@@ -12,3 +12,243 @@
+            {
+                "pdf_path": str(pdf_path),
+                "document_id": str(document_id) if document_id else None,
+                "pages": pages,
+                "extract_text": extract_text,
+                "workflow_step": "marker_layout_extraction",
+            }
+        )
+        current_job.save_meta()
+
+    try:
+        pdf_path = Path(pdf_path)
+        if not pdf_path.exists():
+            raise FileNotFoundError(f"PDF file not found: {pdf_path}")
+
+        _LOGGER.info(f"Starting Marker layout extraction for: {pdf_path}")
+
+        # Call Marker's layout detection
+        # Use the real PdfConverter API for layout detection
+        try:
+            # Process the PDF and get layout information
+            layout_result = _call_marker_layout_detection(str(pdf_path), pages)
+        except Exception as e:
```
Currently, passing in the `pdf_path` is fine for now as it makes it easier to test/develop, but later on it'll need to be converted to using `s3_url` to download the pdf as bytes, then pass it to `marker` as bytes and not a local file path

priyankeshh committed:
Refactor marker integration to use JSONOutput and optimize for layout detection

- Implement three-function structure: create_marker_config(), run_marker(), parse_marker_output()
- Fix marker to return JSONOutput instead of MarkdownOutput using ConfigParser
- Optimize configuration for layout detection workflow
- Successfully extract layout data with proper JSON structure

JonnyTran committed:
refactoring

JonnyTran committed:
refactoring

JonnyTran committed:
Merge branch 'develop' into feat/marker-layout-extraction

JonnyTran committed:
updated typer>=0.19.0
